### PR TITLE
New detox test command to run test cases when device is locked, one f…

### PR DESCRIPTION
**Pull request by varungupta85**
…ailing test case for Android when app is started by tapping on a notification and device is locked and the app is not running in the background.

The locked test cases are run on **Nexus 5 API 25 emulator** which is locked using a **1234 PIN**. The command to run the test cases when  the device is locked is `npm run test-e2e-android-locked`

The command will run two testcases out of which one will fail. The failing testcases reproduces the behavior when user tries to tap on a local notification from the app and the device is locked and the app is not running in the background. In such cases the app gets stuck which is not expected.

The testcases are added to showcase issue https://github.com/wix/react-native-navigation/issues/4675

### CODE REVIEW

File: e2e/DeviceLocked.test.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const KEY_CODE_R = 46;
+
+describe('application lifecycle test', () => {
+  beforeEach(async () => {
+    await device.relaunchApp();
+  });
+
+  test('launch from locked screen', async () => {
+    await device.terminateApp();
+    await Android.pressKeyCode('26')
```
Comment (guyca) [inline e2e/DeviceLocked.test.js]: What does `Android.pressKeyCode('26')` do?

File: e2e/DeviceLocked.test.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    await Android.pressKeyCode('26')
+    await device.launchApp();
+    await Android.executeShellCommand(`input touchscreen swipe 300 1200 500 0 100`)
```
Comment (guyca) [inline e2e/DeviceLocked.test.js]: What does this line do? I assume it's used to swipe across the screen to open the device, if so lets extract it to a function with a meaningful name

File: e2e/DeviceLocked.test.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+describe('application lifecycle test', () => {
+  beforeEach(async () => {
+    await device.relaunchApp();
+  });
+
+  test('launch from locked screen', async () => {
+    await device.terminateApp();
+    await Android.pressKeyCode('26')
+    await Android.pressKeyCode('26')
```
Comment (guyca) [inline e2e/DeviceLocked.test.js]: Why are we invoking this twice?

File: package.json [hunk 4]
```diff
@@ -32,6 +32,7 @@
     "test-unit-android": "node ./scripts/test-unit --android",
     "pretest-e2e-android": "npm run build",
     "test-e2e-android": "node ./scripts/test-e2e --android",
+    "test-e2e-android-locked": "node ./scripts/test-e2e --android --locked",
```
Comment (guyca) [inline package.json]: At its current state, this PR requires a dedicated emulator and for the user to manually setup lock screen with `1234` as password.
Until we can automate this (Probably with UiAutomator) I'd like to remove it from `test-e2e.js` script and into a dedicated script.

File: scripts/test-e2e.js [hunk 5]
```diff
@@ -6,18 +6,21 @@ const release = _.includes(process.argv, '--release');
 const skipBuild = _.includes(process.argv, '--skipBuild');
 const headless = _.includes(process.argv, '--headless');
 const multi = _.includes(process.argv, '--multi');
+const locked = _.includes(process.argv, '--locked');
```
Comment (guyca) [inline scripts/test-e2e.js]: Lets keep this file unchanged and move all logic to a dedicated script.

File: e2e/AndroidUtils.js [hunk 6]
```diff
@@ -21,5 +21,8 @@ module.exports = {
   },
   executeShellCommand: (command) => {
     exec.execSync(`adb shell ${command}`);
+  },
+  unlockPhoneByPin: (pin) => {
+    exec.execSync(`adb shell input text ${pin} && adb shell input keyevent 66`);
```
[the remark this answers is not in the document]
Comment (varungupta85) [reply e2e/AndroidUtils.js]: key event `66` presses the enter button that is required to enter the pin. I will rename it `enterText` and call `doWhatever` as `pressEnter`

File: e2e/DeviceLocked.test.js (hunk 1 above)
Comment (varungupta85) [reply e2e/DeviceLocked.test.js]: `Android.pressKeyCode('26')` is to press the lock button to lock the phone.

File: e2e/DeviceLocked.test.js (hunk 3 above)
Comment (varungupta85) [reply e2e/DeviceLocked.test.js]: One is to lock the device and the other is to press it again to start the unlock process.

File: e2e/DeviceLocked.test.js (hunk 2 above)
Comment (varungupta85) [reply e2e/DeviceLocked.test.js]: yes, the `input touchscreen` command it to perform a swipe on the device such that the device asks the PIN to unlock it.

**Comment by varungupta85** [comment]
hi @guyca, thanks for taking the time to review the PR. I have made all the required changes. I have added a new script `test-e2e-android-locked` that contains the new configuration. Please take a look and let me know if further changes are required. One thing I noted that using the latest `master` branch, I don't see any output after I run the detox test. When I run the `npm run test-e2e-android`, I see below output and no errors.

```
> react-native-navigation@2.20.2 pretest-e2e-android /Users/varungupta/Projects/react-native-navigation
> npm run build


> react-native-navigation@2.20.2 build /Users/varungupta/Projects/react-native-navigation
> rm -rf ./lib/dist && tsc

Files:           111
Lines:         90961
Nodes:        292677
Identifiers:  101860
Symbols:       66055
Types:          7066
Memory used: 117325K
I/O read:      0.06s
I/O write:     0.05s
Parse time:    0.97s
Bind time:     0.37s
Check time:    0.65s
Emit time:     0.30s
Total time:    2.29s

> react-native-navigation@2.20.2 test-e2e-android /Users/varungupta/Projects/react-native-navigation
> node ./scripts/test-e2e --android

detox build --configuration android.emu.debug
detox test --configuration android.emu.debug
```

I see similar output when I run `npm run test-e2e-android-locked` also. I don't know if the instructions to run the detox tests have changed in the latest master.

File: e2e/DeviceLocked.test.js (hunk 1 above)
Comment (guyca) [reply e2e/DeviceLocked.test.js]: Then lets extract this line to a method in `AndroidUtils.js` -`pressLockButton()`

File: e2e/DeviceLocked.test.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+const Utils = require('./Utils');
+const Android = require('./AndroidUtils');
+
+const { elementByLabel } = Utils;
+
+describe('application lifecycle test', () => {
```
Comment (guyca) [inline e2e/DeviceLocked.test.js]: Forgot to rename the describe :)

File: e2e/AndroidUtils.js [hunk 8]
```diff
@@ -21,5 +24,12 @@ module.exports = {
   },
   executeShellCommand: (command) => {
     exec.execSync(`adb shell ${command}`);
+  },
+  unlockPhoneByPin: (pin) => {
```
Comment (guyca) [inline e2e/AndroidUtils.js]: The name is misleading - the method enters text to the currently focused InputText and presses enter - for simplicity it should enter text and that's it.

1. Please rename it it `enterText(text)`
2. It should do one thing, enter text. In your test, call `enterText` following by `pressEnter`

**Comment by varungupta85** [comment]
hi @guyca, I have made the changes that you have requested. I have changed the description to `Android phone locked tests`. I don't know if you have something else in mind. If you do, then please let me know and I can change it to that. Thanks!

**Pull request merged**